### PR TITLE
Add CalmPropertyDataset & CalmLinearProbe Callbacks

### PR DESCRIPTION
**Pull request by taylormjs**
- Add CalmPropertyDataset, downloading from github repo and Zenodo (after putting GO data on Zenodo)
- Add CalmLinearProbeCallback, which randomly subsamples large dataset sizes to 3,000
- Add tests for both of these
- Update LinearProbeCallback to handle multilabel tasks -- Calm has four multilabel tasks (function & localization)
- Update __init__s accordingly

### CODE REVIEW

File: src/lobster/callbacks/_calm_linear_probe_callback.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        self,
+        max_length: int,
+        tasks: Optional[Sequence[str]] = None,
+        species: Optional[Sequence[str]] = None,
+        batch_size: int = 32,
+        run_every_n_epochs: Optional[int] = None,
+        test_size: float = 0.2,
+        max_samples: int = 3000,
+        seed: int = 42,
+    ):
```
Comment (ncfrey) [inline src/lobster/callbacks/_calm_linear_probe_callback.py:30]: can you add a docstring, especially for explaining `tasks` and `species` ?

File: src/lobster/callbacks/_calm_linear_probe_callback.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        self.dataset_splits = {}
+        self.aggregate_metrics = defaultdict(list)
+
+    def _create_split_datasets(
+        self, task: str, species: Optional[str] = None
+    ) -> Tuple[CalmPropertyDataset, CalmPropertyDataset]:
+        """Create train/test splits for a given task."""
+
+        rng = np.random.RandomState(self.seed)  # TODO - seed everything fn
```
Comment (ncfrey) [inline src/lobster/callbacks/_calm_linear_probe_callback.py:60]: hmm, does this work better or equivalently to lightning's seed everything?

File: src/lobster/callbacks/_linear_probe_callback.py [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 from beignet.transforms import Transform
 from lightning.pytorch.callbacks import Callback
 from sklearn.linear_model import LinearRegression, LogisticRegression
+from sklearn.multioutput import MultiOutputClassifier
 from torch import Tensor
 from torch.utils.data import DataLoader
 from torchmetrics import AUROC, Accuracy, F1Score, MeanSquaredError, R2Score, SpearmanCorrCoef
 
-TaskType = Literal["regression", "binary", "multiclass"]
+TaskType = Literal["regression", "binary", "multiclass", "multilabel"]
```
Comment (ncfrey) [inline src/lobster/callbacks/_linear_probe_callback.py:14]: is multilabel different from multiclass?

File: src/lobster/constants/_calm_tasks.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    "transcript_abundance": ("regression", None),
+    "function_bp": ("multilabel", 4),  # 4 GO terms
+    "function_cc": ("multilabel", 4),
+    "function_mf": ("multilabel", 4),
+}
+
+
+CALM_DATA_GITHUB_URL = "https://raw.githubusercontent.com/oxpig/CaLM/main/data"
+FUNCTION_ZENODO_BASE_URL = (
+    "https://zenodo.org/records/14890750/files"  # Gene Ontology datasets processed & uploaded on Zenodo
```
Comment (ncfrey) [inline src/lobster/constants/_calm_tasks.py:17]: switch to HF

File: src/lobster/datasets/_calm_property_dataset.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        if str(file_path).endswith(".parquet"):
+            self.data = pd.read_parquet(file_path)
+        elif str(file_path).endswith(".tsv"):
+            self.data = pd.read_csv(file_path, sep="\t")
+        else:
+            self.data = pd.read_csv(file_path)
+
+        if columns is None:
+            if task == Task.FUNCTION_BP:
```
Comment (ncfrey) [inline src/lobster/datasets/_calm_property_dataset.py:101]: maybe a match case here to handle the logic?

File: src/lobster/model/modern_bert/_modern_bert.py [hunk 6]
```diff
@@ -50,46 +43,18 @@ def __init__(
         self._mask_percentage = mask_percentage
         self.max_length = max_length
 
-        # TODO zadorozk: currently only accepts one tokenizer at a time
```
Comment (ncfrey) [inline src/lobster/model/modern_bert/_modern_bert.py:53]: undo delete

File: src/lobster/constants/_calm_tasks.py (hunk 4 above)
Comment (karinazad) [reply src/lobster/constants/_calm_tasks.py:17]: I was going to write something similar about CALM_DATA_GITHUB_URL -- is this a subset of the calm dataset from Nathan's HF account?

Comment (karinazad) [reply src/lobster/constants/_calm_tasks.py:17]: if it's in HF, we can also load it more easily with `datahaha` once we switch

File: src/lobster/constants/_calm_tasks.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        Species.DMELANOGASTER,
+        Species.ECOLI,
+        Species.HSAPIENS,
+        Species.HVOLCANII,
+        Species.PPASTORIS,
+        Species.SCEREVISIAE,
+    ],
+}
+
+# Files hashes to check upstream data files haven't been changed. Makes data download cleaner
```
Comment (karinazad) [inline src/lobster/constants/_calm_tasks.py:61]: One thing I don't love about hashes is that pooch currently just errors out when the files are different. Maybe we'd want to add custom logic where it just downloads new files instead. Could be something for a future MR though

File: src/lobster/datasets/_calm_property_dataset.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
+from pathlib import Path
+from typing import Callable, Literal, Optional, Sequence, Tuple
+
+import pandas as pd
+import pooch
+import torch
+from torch import Tensor
+from torch.utils.data import Dataset
+
+from lobster.constants._calm_tasks import (
```
Comment (karinazad) [inline src/lobster/datasets/_calm_property_dataset.py:10]: generally good not to import from modules with an underscore. If they are in constants. init, we can just do from `lobster.constants`

File: src/lobster/model/modern_bert/_modern_bert.py [hunk 9]
```diff
@@ -35,8 +28,8 @@ def __init__(
         eps: float = 1e-12,
         num_training_steps: int = 10_000,
         num_warmup_steps: int = 1_000,
-        tokenizer: Union[str, PreTrainedTokenizer, PreTrainedTokenizerFast] = "amino_acid_tokenizer",
```
Comment (karinazad) [inline src/lobster/model/modern_bert/_modern_bert.py:38]: why was this deleted?

File: src/lobster/model/modern_bert/_modern_bert.py [hunk 10]
```diff
@@ -50,46 +43,18 @@ def __init__(
...
-        # TODO zadorozk: currently only accepts one tokenizer at a time
-        # Extend to accept multiple tokenizers for each modality
-        if isinstance(tokenizer, str):
-            if tokenizer == "pmlm_tokenizer":
-                path = importlib.resources.files("lobster") / "assets" / "plm_tokenizer"
-                tokenizer = PmlmTokenizer.from_pretrained(path, do_lower_case=False)
-                tokenizer_transform_class = PmlmTokenizerTransform
-
-            elif tokenizer == "amino_acid_tokenizer":
-                tokenizer = AminoAcidTokenizerFast()
```
Comment (karinazad) [inline src/lobster/model/modern_bert/_modern_bert.py:62]: we could probably get rid of `tokenizer` inside the model altogether (and also sequence_to_latents?) as it's not needed for training

what do you think @ncfrey 

otherwise we should probably keep the option to provide a custom tokenizer

File: tests/lobster/callbacks/test__calm_linear_probe_callback.py
Comment (karinazad) [inline tests/lobster/callbacks/test__calm_linear_probe_callback.py:1]: nice!

File: src/lobster/callbacks/_calm_linear_probe_callback.py (hunk 2 above)
Comment (taylormjs) [reply src/lobster/callbacks/_calm_linear_probe_callback.py:60]: Good question, I'm guessing lightning's seed everything is more general. I'll switch to that

File: src/lobster/callbacks/_linear_probe_callback.py (hunk 3 above)
Comment (taylormjs) [reply src/lobster/callbacks/_linear_probe_callback.py:14]: In this case, yes. e.g. for cellular localization, a sequence could be found in several of 10 possible cellular compartments (multi-label), whereas I interpreted multiclass as belonging to exactly one of many classes

File: src/lobster/constants/_calm_tasks.py (hunk 4 above)
Comment (taylormjs) [reply src/lobster/constants/_calm_tasks.py:17]: The data at CALM_DATA_GITHUB_URL is an entirely separate dataset, not a subset of hf/ncfrey/calm. It's also different than the calm validation set used for pre-training. Perhaps I could have a property_pred dataset that also lives in Nathan's hf account? Could also put in my own

File: src/lobster/constants/_calm_tasks.py (hunk 7 above)
Comment (taylormjs) [reply src/lobster/constants/_calm_tasks.py:61]: Oh interesting, yeah that's annoying. I can add that in the next MR

File: src/lobster/datasets/_calm_property_dataset.py (hunk 5 above)
Comment (taylormjs) [reply src/lobster/datasets/_calm_property_dataset.py:101]: Yep, definitely. Will change

File: src/lobster/model/modern_bert/_modern_bert.py (hunk 9 above)
Comment (taylormjs) [reply src/lobster/model/modern_bert/_modern_bert.py:38]: Ah, I may have rebased on the wrong commit. Will re-add this

File: src/lobster/model/modern_bert/_modern_bert.py (hunk 10 above)
Comment (ncfrey) [reply src/lobster/model/modern_bert/_modern_bert.py:62]: @karinazad agreed!

**Review by ncfrey — APPROVED** [review]
